### PR TITLE
[[FIX]] parse `const` declarations in ForIn/Of loops

### LINKED ISSUE #2334: Fails to properly parse for..of loops with const bindings
The following is valid ES6 code that prevents accidental assignments to `x` inside of the loop body:

```js
/*jshint esnext:true, node:true*/
'use strict';
for (const x of [ 1, 2, 3 ]) { console.log('x', x); }
```

Running with io.js 1.8:

```
> node -e "/*jshint esnext:true, node:true*/ 'use strict'; for (const x of [ 1, 2, 3 ]) { console.log('x', x); }"
x 1
x 2
x 3
```

But jshint doesn't seem to expect a `const` token in that position, interpreting it as an identifier:

```
> jshint <(echo "/*jshint esnext:true, node:true*/ 'use strict'; for (const x of [ 1, 2, 3 ]) { console.log(x); }") 
/dev/fd/11: line 1, col 54, Creating global 'for' variable. Should be 'for (var const ...'.
/dev/fd/11: line 1, col 60, Expected 'of' and instead saw 'x'.
/dev/fd/11: line 1, col 68, Expected ']' to match '[' from line 1 and instead saw ','.
/dev/fd/11: line 1, col 70, Expected ')' to match '(' from line 1 and instead saw '2'.
/dev/fd/11: line 1, col 71, Expected an identifier and instead saw ','.
/dev/fd/11: line 1, col 71, Expected an assignment or function call and instead saw an expression.
/dev/fd/11: line 1, col 72, Missing semicolon.
/dev/fd/11: line 1, col 73, Expected an assignment or function call and instead saw an expression.
/dev/fd/11: line 1, col 74, Missing semicolon.
/dev/fd/11: line 1, col 75, Expected an identifier and instead saw ']'.
/dev/fd/11: line 1, col 75, Expected an assignment or function call and instead saw an expression.
/dev/fd/11: line 1, col 76, Missing semicolon.
/dev/fd/11: line 1, col 76, Expected an identifier and instead saw ')'.
/dev/fd/11: line 1, col 76, Expected an assignment or function call and instead saw an expression.
/dev/fd/11: line 1, col 77, Missing semicolon.
/dev/fd/11: line 1, col 62, 'of' is not defined.
/dev/fd/11: line 1, col 92, 'x' is not defined.

17 errors
```

The same code does work as expected with `var` or `let` instead of `const`.

### PR DESCRIPTION
**Pull request by caitp**
Closes #2334

### CODE REVIEW

File: src/jshint.js [hunk 1]
```diff
@@ -4189,8 +4189,8 @@ var JSHINT = (function() {
       if (state.tokens.next.id === "var") {
         advance("var");
         state.tokens.curr.fud({ prefix: true });
-      } else if (state.tokens.next.id === "let") {
-        advance("let");
+      } else if (state.tokens.next.id === "let" || state.tokens.next.id === "const") {
```
Comment (caitp) [inline src/jshint.js]: TODO: should only be treating let/const specially if `esnext` es3 or es5 mode is used

File: src/jshint.js [hunk 2]
```diff
@@ -4189,8 +4189,8 @@ var JSHINT = (function() {
...
+        advance(state.tokens.next.id);
```
Comment (caitp) [inline src/jshint.js]: other linting rules: initializers are illegal in es6 here, as are multiple declarations

**Comment by caitp** [comment]
self-reviewed, think it should wait on fixing the behaviour of `const` though.

**Comment by rwaldron** [comment]
I'm mobile atm so can't compare with https://github.com/jshint/jshint/pull/2326 is there any overlap? (Sorry if that seems lazy)

**Comment by caitp** [comment]
not that I can see in the diff

**Comment by caitp** [comment]
it looks like coveralls.io is having issues

**Comment by lukeapage** [comment]
I don't think it conflicts with my pr :)

**Comment by caitp** [comment]
it's not conflicts that's the worry, it's that the behaviour of `const` is pretty broken atm =)

**Comment by lukeapage** [comment]
Im working on it - i have work i havent pushed yet as well.  If you can
review my pr (im building on top of it so sooner the better for code
review) and get them both in,  that would help and i can create my next pr.

If you know of anything const specific (next pr is for latedef and unused
of block scope,  so let too) thats very broken and not fixed by this and my
pr,  please let me know.

**Pull request closed without merging**